[DRAFT] Mob Container Module

NotAGhostWriter opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds a new component, `mob_container`, that provides functionality for storing a mob safely inside of another mob. Breathing is done on the turf of the mob you're inside of. This currently does not add anything to the user experience, as I'll leave that up to someone else. You know what I mean.

For custom functionality, you may create a subtype of `/datum/component/mob_container`, and alter the 4 given procs at the very end of the file. These procs are ran every single time someone enters/exit a mob's container. You can forcibly insert or remove a mob by using `SEND_SIGNAL(mob, COMSIG_CONTAINER_STORE/REMOVE_MOB`, mob_to_store/remove)`.

Notice: This PR is being made on a ghost-writer account, due to the.. controversial nature of this PR. Attempting to find who the original author of this code is is a major dick move.

## How This Contributes To The Skyrat Roleplay Experience

Host request.

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

Jolly-66 commented:
If the PR is actually a draft, please click the "Convert to Draft' button under reviewers.

NotAGhostWriter commented:
> If the PR is actually a draft, please click the "Convert to Draft' button under reviewers.

Woops.

theOOZ commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,36 @@
+	if(!tostore.on_enter_container(src) || !on_container_entered(tostore))
+		return
+
+	ADD_TRAIT(tostore, TRAIT_OXYIMMUNE, REF(src)) // We give the mob inside oxy-damage immunity. Hope this doesn't come back to bite us in the ass.
```
:/ power crep

softcerv replied:
extreme role play mechanics don't need balance.

tf-4 replied:
yes they do

Zonespace27 replied:
> extreme role play mechanics don't need balance.

Nah mate we don't need shit like the leather whip being a worse police baton

softcerv replied:
> > extreme role play mechanics don't need balance.
> 
> Nah mate we don't need shit like the leather whip being a worse police baton

Doesn't that make it far easier to single out bad actors tho?

Iamgoofball replied:
no

Ghommie commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,36 @@
+	if(!tostore.on_enter_container(src) || !on_container_entered(tostore))
+		return
+
+	ADD_TRAIT(tostore, TRAIT_OXYIMMUNE, REF(src)) // We give the mob inside oxy-damage immunity. Hope this doesn't come back to bite us in the ass.
```
Can we explain why the stored mob is given immunity to such damage in the comment? Also can you change it to TRAIT_NOBREATH if it's simply an issue with the lack of breathable air.

Iamgoofball replied:
you should instead just redirect the breath out to the tile the mob is on

Ghommie replied:
For once, I agree. That'd be a good idea.

Ghommie commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,36 @@
+	if(!toremove.on_exit_container(src) || !on_container_exited(toremove))
+		return
+
+	REMOVE_TRAIT(toremove, TRAIT_OXYIMMUNE, REF(src)) // Remove trait prevents the loss of roundstart traits, and we also have ourselves as a source. So don't worry.
```
ditto on the no breath trait.

tf-4 commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,36 @@
+/mob/proc/store_mob(mob/tostore)
```
does this need to be on the mob level? I'd rather it were on the stomach.

Gandalf2k15 reviewed (changes requested):
no no no no no make this a component oh god

Iamgoofball commented:
This in and of itself isn't controversial, and would actually be useful upstream.

Zonespace27 reviewed (changes requested):
Componentize this and it should very likely be on the stomach, not the mob

Iamgoofball commented:
Probably shouldn't make it on the stomach, arguably I think the better route here might be refactoring storage to be able to contain all types of atoms instead of just /obj/item.

NotAGhostWriter commented:
This won't be going onto the stomach: as this PR has no intent to make it related to the stomach whatsoever. This is simply a way for mobs to contain other mobs. However, it is currently being componentized.

Zenitheevee commented:
> Probably shouldn't make it on the stomach, arguably I think the better route here might be refactoring storage to be able to contain all types of atoms instead of just /obj/item.

"No, Mr Antag, My duffelbag does not contain 20 armed teshari, now i'm just going to leave it here in the corner of the room and go take a wizz. Do not open it please it contains all my confiscated contraband items."

Iamgoofball commented:
exactly

YakumoChen commented:
The big question becomes how expensive is this?

Last time we rejected ""mob engulfing"" because it was insanely unoptimised.

Gandalf2k15 commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,94 @@
+	if(!istype(mob_holder))
+		return COMPONENT_INCOMPATIBLE
+
+	src.mob_holder = WEAKREF(mob_holder)
```
two things, don't use src. and secondly, components get added to something and the parent becomes that thing, don't have the thing it's being added to passed in on initialise

Zonespace27 commented on code/modules/mob/living/carbon/life.dm:
```diff
@@ -74,6 +74,11 @@
 			var/obj/location_as_object = loc
 			location_as_object.handle_internal_lifeform(src,0)
+		// SKYRAT EDIT BEGIN - MOB CONTAINER
+		if(istype(loc, /mob/))
```
`ismob()`

Zonespace27 commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,94 @@
+
+	UnregisterSignal(mob_holder, list(COMSIG_MOB_CONTAINER_ENTERED, COMSIG_MOB_CONTAINER_EXITED, COMSIG_MOB_ENTER_CONTAINER, COMSIG_MOB_EXIT_CONTAINER))
+
+/datum/component/mob_container/proc/mob_breathe_check()
```
code doc

Zonespace27 commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,94 @@
+/datum/component/mob_container/proc/mob_breathe_check()
+	return container_breathable
+
+/datum/component/mob_container/proc/store_mob_explicitly(mob/to_store)
```
ditto

Zonespace27 commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,94 @@
+	to_store.overlay_fullscreen("in_mob_container", fullscreen_overlay, 1)
+	mobs_held |= to_store
+
+/datum/component/mob_container/proc/remove_mob_explicitly(mob/to_remove)
```
ditto

Zonespace27 commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,94 @@
+#define COMSIG_MOB_CONTAINER_ENTERED "mob_container_entered"
```
don't define signals here

NotAGhostWriter replied:
These will be moved. Everything is being kept in one place for now since it's a draft.

softcerv commented:
> The big question becomes how expensive is this?
> 
> Last time we rejected ""mob engulfing"" because it was insanely unoptimised.

We removed it because we rebased and nobody ended up coding it back in. This doesn't look to be a port of that system.

tf-4 commented:
> > The big question becomes how expensive is this?
> > Last time we rejected ""mob engulfing"" because it was insanely unoptimised.
> 
> We removed it because we rebased and nobody ended up coding it back in. This doesn't look to be a port of the system from oldbase.

It not being an oldbase port doesn't inherently make it good code. Not saying it isn't, even, but it's a valid question.

NotAGhostWriter commented:
> > > The big question becomes how expensive is this?
> > > Last time we rejected ""mob engulfing"" because it was insanely unoptimised.
> > 
> > 
> > We removed it because we rebased and nobody ended up coding it back in. This doesn't look to be a port of the system from oldbase.
> 
> It not being an oldbase port doesn't inherently make it good code. Not saying it isn't, even, but it's a valid question.

This will have no noticeable impact on performance. Essentially, all we're doing is forcibly moving this mob into another mob's contents, and applying a fullscreen overlay. Likely the most performance intensive addition will be a subversion inside `breathe()`, which checks if the user is inside of a mob instead of a turf.

I'll upload a profile after I get the final version.

Ghommie commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,177 @@
+ *
+ * mob/mob_were_exiting | the mob we're being taken out of
+ */
+/datum/component/mob_container/proc/on_exit_container(datum/source, mob/mob_were_exiting)
```
Any good reason this ought to be called by sending a COMSIG_MOB_CONTAINER_EXITED signal rather than doing so right away on line 132? Don't get me wrong, the signals are ok, but calling "native" procs this way is pretty convoluted, inefficient.
Also, since you're adding so signals already, perhaps you should ditch the idea of having subtypes of this component with their own quirks. Let it be done by other datums, components or elements and keep it simple here.

NotAGhostWriter replied:
Meh, it isn't a huge deal. Efficiency wise, it isn't too terribly different from calling it natively (unless the load is large), however I think this is something worth changing.

Component subtypes aren't a new thing by any means. I'm going to leave it on the subtype idea, that can be changed in the future if absolutely dire.

NotAGhostWriter replied:
Meh, it isn't a huge deal. Efficiency wise, it isn't too terribly different from calling it natively (unless the load is large), however I think this is something worth changing.

Component subtypes aren't a new thing by any means. I'm going to leave it on the subtype idea, that can be changed in the future if absolutely dire.

Ghommie commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,177 @@
+/datum/component/mob_container/proc/dump_all_mobs()
+	SIGNAL_HANDLER
+
+	for(var/mob/contained in mobs_held)
```
```suggestion
	for(var/mob/contained as anything in mobs_held)
```
[Avoid unnecessary type checks and obscuring nulls in lists](https://github.com/Skyrat-SS13/Skyrat-tg/blob/master/.github/guides/STANDARDS.md#avoid-unnecessary-type-checks-and-obscuring-nulls-in-lists=)

NotAGhostWriter commented on modular_skyrat/modules/mob_container/mob_containers.md:
```diff
@@ -0,0 +1,43 @@
+Mob containers are for the express purpose of storing a mob inside of another mob. This is not for antagonistic nor further mechanical purposes, and should not be used for such things. This is not a solution for having atomic storage (see /datum/storage) inside of mobs: this is only for mob inside mob storage.
+# Usage
+## Adding a Mob to a Mob
+Adding a mob to another mob is easy! Simply send the signal as following:
```
I have no idea where you possibly could've heard this. From my experience, signals are not a "one-use" sort of deal. It's a signal to specify an action.

NotAGhostWriter commented on modular_skyrat/modules/mob_container/mob_container.dm:
```diff
@@ -0,0 +1,177 @@
+
+/datum/component/mob_container
+	/// The mob we're attached to.
+	var/mob/mob_holder
```
This exists only to avoid re-pathing the parent every single time it's called. The performance difference is negligible.

Gandalf2k15 reviewed (changes requested):
Signal abuse is not allowed, do not abuse signals like this.
